### DISCUSSION TITLE
AWS IoT MQTT issues

### DISCUSSION THREAD

**Discussion by bertmelis** [bertmelis/espMqttClient#121] [General]
AWS has a strict implementation of the MQTT standard. Although both AWS and espMqttClient follow the same standard and are largely compliant, care should be taken when implementing. A number of points in specific:

- AWS doesn't support QoS 2. Publishing, setting a last will and (possibly?) subscribing with QoS will result in a disconnect. MQTT 3.1.1 doesn't support specific disconnect messages so AWS just closes the TCP connection without further information to the client.
**Do not use QoS 2 for AWS!**
- AWS doesn't support publishing before a complete MQTT connection is established. Make sure you have `EMC_WAIT_FOR_CONNACK` set to `1`, which is also the default. **Leave this default as is.**
- AWS doesn't guarantee the order of ACKs. espMqttClient expects the ACKs to come in in the same order the PUBLISH messages are sent, duplicates may be sent for out of order received ACKs. Message delivery retry is done by espMqttClient automatically. **Make sure your application can handle duplicate messages and message delivery retries.**

These are the most obvious things to watch out for. There may be other issues. A full list of AWS implementation specifics can be found here: https://docs.aws.amazon.com/iot/latest/developerguide/mqtt.html#mqtt-differences

**Comment by proddy** [comment]
semi-related, there are setup guides at https://github.com/philipparndt/mqtt-analyzer/tree/main/Docs/examples/aws for reference, maybe, don't know, just want to join the fun ;)
